play-services-fido/core/src/main/kotlin/org/microg/gms/fido/core/transport/TransportHandler.kt

```diff
@@ -156,20 +199,50 @@ abstract class TransportHandler(val transport: Transport, val callback: Transpor
         connection: CtapConnection,
         context: Context,
         options: RequestOptions,
-        callerPackage: String
+        callerPackage: String,
+        pinRequested: Boolean,
+        pin: String?
     ): AuthenticatorAttestationResponse {
         val (clientData, clientDataHash) = getClientDataAndHash(context, options, callerPackage)
         val (response, keyHandle) = when {
             connection.hasCtap2Support -> {
-                if (connection.hasCtap1Support &&
-                    !connection.canMakeCredentialWithoutUserVerification && connection.hasClientPin &&
-                    options.registerOptions.authenticatorSelection?.requireUserVerification != REQUIRED &&
-                    options.registerOptions.authenticatorSelection?.requireResidentKey != true
-                ) {
-                    Log.d(TAG, "Using CTAP1/U2F for PIN-less registration")
+                if (connection.hasCtap1Support
+                    && options.registerOptions.authenticatorSelection?.residentKeyRequirement == RESIDENT_KEY_DISCOURAGED) {
```

Comment:
If I've understood CTAP1 correctly, it can do a lot of the same things as CTAP2, but it cannot handle resident keys or PIN authentication. If I'm not mistaken, user verification works in CTAP1 if the verification is built in to the authenticator. The code is now written to default to CTAP1 when registering unless resident keys and/or PIN authentication is required.